### PR TITLE
Abstract transport

### PR DESCRIPTION
**Pull request by hristoterezov**
https://github.com/jitsi/jitsi-meet-electron/pull/7/files should be merged with this PR.

### CODE REVIEW

**Comment by saghul** [comment]
@hristoterezov I won't comment on the code itself, but the design.

In its current implementation, this is not something we can reuse, I'm afraid. The API uses some transport instance which is global, this is something we are trying to reduce, not add to.

I think the relationship between API, Transport and PostMessageTransport is weird. The way I would envision it is as follows:

API uses a transport, which gets in the constructor. API is the entity responsible for processing incoming commands from the transport, creating actions and sending responses to the commands and events through the transport.

The Transport would have a simple API for the API class to consume:
- a send function
- a single event emitted: dataReceived

It's the responsibility of API to process these commands.

Right now the implementation piggybacks on how postis works for sending stuff and listening for events. I think we should only send 1 type of message to postis, and the command / response / event would be the payload, postis is just a transport, ignorant of the data being carried. Same goes for listening for events, there should just be one event for data reception, postis shouldn't be the entity doing the command handling, just transport.

In addition, if we are to reuse this we need it to be in React.

Last, I guess some of the design choices here were due to backwards compatibility desires, but that is really hurting now, I think we need to reconsider sooner rather than later, and by reconsider I mean break it for once and have something future-proof.

I'm happy to discuss this further, since I probably haven't expressed myself as clear as it is in my head.

**Comment by hristoterezov** [comment]
@saghul Let's discuss it offline. This seems like too long topic for writing :)

File: modules/remotecontrol/Receiver.js [hunk 1]
```diff
@@ -174,6 +185,36 @@ export default class Receiver extends RemoteControlParticipant {
...
+                break;
+        }
+    }
+
+    /**
+     * Handles events for support for executing remote control events into
+     * the wrapper application.
+     */
+    _onRemoteControlSupported() {
+        logger.log("Remote Control supported.");
```
Comment (saghul) [inline modules/remotecontrol/Receiver.js]: style: use single quotes

File: modules/remotecontrol/Receiver.js [hunk 2]
```diff
@@ -174,6 +185,36 @@ export default class Receiver extends RemoteControlParticipant {
...
+        if(!config.disableRemoteControl) {
```
Comment (saghul) [inline modules/remotecontrol/Receiver.js]: style: space after if

File: modules/transport/Transport.js [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    constructor(options = {}) {
+        super();
+
+        const { defaultTransportOptions, noDefaultTransport, transport }
+            = options;
+
+        // We need this because addListener in overriden
+        this.on = this.addListener;
+
+        this._requestID = 0;
```
Comment (hristoterezov) [inline modules/transport/Transport.js]: Yes. Good catch! 

My idea was to increase this on every request but obviously I forgot this :) I'll fix it!

File: modules/transport/Transport.js [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+     *
+     * @param {string} name - The name of the request.
+     * @param {Object} data - The data for the request.
+     * @returns {Promise}
+     */
+    sendRequest(name, data) {
+        if (!this._transport) {
+            return Promise.reject(new Error('No transport defined!'));
+        }
+        const id = this._requestID;
```
Comment (hristoterezov) [inline modules/transport/Transport.js]: Hmm. My idea was to increase this on every request. 

I have to check your proposal. But isn't the increasing by 1 good enough for the initial implementation?

**Comment by hristoterezov** [comment]
About all code style issues for remote control module:
Currently we haven't enabled our eslint config for this module. We can enable it and fix all the eslint errors there. I prefer to do this on separate PR.

File: modules/remotecontrol/Receiver.js (hunk 2 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: See https://github.com/jitsi/jitsi-meet/pull/1510#issuecomment-296710198

File: modules/remotecontrol/Receiver.js (hunk 1 above)
Comment (hristoterezov) [inline modules/remotecontrol/Receiver.js]: See https://github.com/jitsi/jitsi-meet/pull/1510#issuecomment-296710198

File: modules/transport/index.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+export const transport = new Transport({ defaultTransportOptions });
+
+/**
+ * Sets the transport to passed transport.
+ *
+ * @param {Object} newTransport - The new transport.
+ * @returns {void}
+ */
+window.useNewExternalTransport = function(newTransport) {
```
Comment (hristoterezov) [inline modules/transport/index.js]: Current implementation of the transport module is not related to the API module. The transport module can be used even without the API module existence. 

Also currently the only use case for this function will be the remote control for external apps that doesn't support postMessage and the remote control module is not using the API module.

So I don't think this function fits to APP.API.

**Comment by hristoterezov** [comment]
> After our discussion the other day I understand why you wanted to introduce the Transport class. My suggestion is to change that to TransportAdapter and for the consumer to create it with the specific transport in the constructor. Plus have a way to change the downstream transport.
> 

I don’t have any preferences about the naming.

Currently you can do `new Transport({transport: instance_of_transport})` - and this way the `Transport` class will use the passed low-level transport. Also there is `setTransport` method to change the current low-level transport that is used.

> More importantly, the API the downstream transport must implement is quite big now, I'd like to see that reduced to just send setDataCallbackanddispose` methods, and do all the event handling in the TransportAdapter class. This will be needed for use in mobile.
> 

I think currently the interface is very similar to what you are describing:
- There should be `dispose` method.
- there should be `send(name, data)`
- and receiving data is supposed to work trough `EventEmitter` events.

The only difference that I see from what you are describing is that we are using events instead of callback which for me is minor difference.

> As for detecting the use of an old API implementer, I'd move that to the PostMessageTransport, which is the only one which ever got that. This way TransportAdapter doesn't have to bother with backwards compatibility.
> 
All backward compatibility parts are currently included in PostMessageTransport. I don't think currently there is something left in the Transport class. Could you specify which part of the code exactly are you referring to?

> Creating the TransportAdapter instance should be done explicitly, not in a global even if you added a way to change it later. As I mentioned, everything should be contained within the app, so I think building it as part of the API initialization is reasonable.
> 
I'm not sure that I understand your idea here! Do you mean that we are creating only one transport instance for jitsi-meet app? If yes - I don't see any reason why we don't want this.

In Jitsi Meet app(not including external_api.js and other possible usages of the Transport class) we are using transport not only for the API module but for the remote control as well and possibly in future for other modules. I'm not aware of any use cases that we will need inside Jitsi Meet more than one different low-level transports used in the same time. I think we'll always want to use one low-level transport for everything (API module and remote control). If I'm correct about this why we would want to create more than one instance of the low-level transport or the Transport class? For me this adds additional complexity without any benefit because every consumer module should create explicitly at least 2 instances (for the Transport class and for the low-level transport). Also for the postMessage transport and possibly for the other transports we don't want 2 postis instances running simultaneously for obvious reasons. It will be tricky there to implement the class without creating additional postis instance or handling more than one postis instances. Also why each consumer should care which is the correct transport to work with. This should be implemented in the transport module.

> Also left some code comments this time!

File: modules/transport/Transport.js [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    MESSAGE_TYPE_EVENT,
+    MESSAGE_TYPE_RESPONSE,
+    MESSAGE_TYPE_REQUEST
+} from './constants';
+import PostMessageTransport from './PostMessageTransport';
+
+/**
+ * Stores the currnet transport that have to be used.
+ */
+export default class Transport extends EventEmitter {
```
Comment (hristoterezov) [inline modules/transport/Transport.js]: I don't have preference about the name. Although right now it is more of a transport manager or transport facade.

File: modules/transport/Transport.js [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        // We need this because addListener in overriden
+        this.on = this.addListener;
+
+        this._requestID = 0;
+
+        if (transport) {
+            this._transport = transport;
+        } else if (!noDefaultTransport) {
+            // Set the default transport.
+            this._transport = new PostMessageTransport(defaultTransportOptions);
```
Comment (hristoterezov) [inline modules/transport/Transport.js]: I'm not sure about the idea with passing the specific transport to the constructor. See https://github.com/jitsi/jitsi-meet/pull/1510#issuecomment-296810379

Although you can do that even now trough the transport option, but this is not the default behavior.

File: modules/transport/index.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ * @type {Object}
+ */
+const defaultTransportOptions = {};
+
+if (typeof API_ID === 'number') {
+    defaultTransportOptions.scope
+        = `jitsi_meet_external_api_${API_ID}`;
+}
+
+export const transport = new Transport({ defaultTransportOptions });
```
Comment (hristoterezov) [inline modules/transport/index.js]: See https://github.com/jitsi/jitsi-meet/pull/1510#issuecomment-296810379

File: modules/transport/index.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import Transport from './Transport';
+
+/**
+ * Option for the default low level transport.
+ *
+ * @type {Object}
+ */
+const defaultTransportOptions = {};
+
+if (typeof API_ID === 'number') {
```
Comment (hristoterezov) [inline modules/transport/index.js]: Hmm. The problem here is that external_api.js uses Transport class as well but need to provide different options.

File: modules/transport/Transport.js (hunk 4 above)
Comment (hristoterezov) [inline modules/transport/Transport.js]: Fixed!

File: modules/transport/Transport.js (hunk 6 above)
Comment (saghul) [inline modules/transport/Transport.js]: TransportWrapper?

File: modules/transport/index.js (hunk 5 above)
Comment (saghul) [inline modules/transport/index.js]: IMHO it totally fits. Remote control is an API we provide, is it not? Whatever the API module meant in the past we can extend it now.

**Comment by saghul** [comment]
> About all code style issues for remote control module:
> Currently we haven't enabled our eslint config for this module. We can enable it and fix all the eslint errors there. I prefer to do this on separate PR.

For the old code, I agree, for new code, I'd rather see it written correctly from the start.

**Pull request closed without merging**